src/helpers/mock.rs

```diff
@@ -136,117 +213,106 @@ impl<S: Step> Mesh for TestMesh<S> {
 impl<S> Clone for Controller<S> {
     fn clone(&self) -> Self {
         Self {
+            receive_request_sender: self.receive_request_sender.clone(),
             identity: self.identity,
             peers: self.peers.clone(),
             connections: Arc::clone(&self.connections),
-            buf: Arc::clone(&self.buf),
         }
     }
 }
 
 impl<S: Step> Controller<S> {
     fn launch(
         identity: Identity,
-        control_tx: HashMap<Identity, Sender<ControlMessage<S>>>,
-        control_rx: Receiver<ControlMessage<S>>,
+        control_tx: HashMap<Identity, mpsc::Sender<ControlMessage<S>>>,
+        control_rx: mpsc::Receiver<ControlMessage<S>>,
     ) -> Self {
+        let (receive_tx, receive_rx) = mpsc::channel(1);
         let controller = Self {
+            receive_request_sender: receive_tx,
             identity,
             connections: Arc::new(Mutex::new(HashMap::new())),
-            buf: Arc::new(Mutex::new(HashMap::new())),
             peers: control_tx,
         };
 
-        controller.start(control_rx);
+        Controller::start(control_rx, receive_rx);
 
         controller
     }
 
-    fn start(&self, mut rx: Receiver<ControlMessage<S>>) {
-        tokio::spawn({
-            let controller = self.clone();
-            async move {
-                while let Some(msg) = rx.recv().await {
-                    match msg {
-                        ControlMessage::ConnectionRequest(peer, step, peer_connection) => {
-                            controller.connect(peer, step, peer_connection);
+    fn start(
+        mut control_rx: mpsc::Receiver<ControlMessage<S>>,
+        mut receive_rx: mpsc::Receiver<ReceiveRequest<S>>,
+    ) {
+        tokio::spawn(async move {
+            let mut buf = HashMap::<ChannelId<S>, MessageBuffer>::new();
+            let mut channels = SelectAll::new();
+
+            loop {
+                // Make a random choice what to process next:
+                // * Receive and process a control message
+                // * Receive a message from another helper
+                // * Handle the request to receive a message from another helper
+                tokio::select! {
+                    Some(control_message) = control_rx.recv() => {
+                        match control_message {
+                            ControlMessage::ConnectionRequest(peer, step, peer_connection) => {
+                                channels.push(prepend((peer, step), ReceiverStream::new(peer_connection)));
```

Comment:
Do you need to implement prepend in the other module?  It seems like a fairly specialized tweak, so maybe it belongs here (at least for now).